providers/shared/fqdn.go

```diff
@@ -0,0 +1,152 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build linux || darwin
+
+package shared
+
+import (
+	"bufio"
+	"fmt"
+	"io/fs"
+	"net"
+	"os"
+	"strings"
+)
+
+const etcHosts = "/etc/hosts"
+
+func FQDN() (string, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", fmt.Errorf("could get hostname to look for FQDN: %w", err)
+	}
+
+	var errs error
+
+	fqdn, err := fqdnFromEtcHosts(hostname)
+	if err != nil {
+		errs = fmt.Errorf("could not get FQDN, all methods failed: %w", err)
+	}
+	if fqdn != "" {
+		return fqdn, nil
+	}
```

Comment:
I think I'm missing something.  Why parse `/etc/hosts`?  The `net.LookupCNAME`, `net.LookuIP` and `net.LookupAddr` will use `/etc/hosts` if the resolver on the system uses that file for name resolution.  Is there a specific use case this is addressing?  If so I really think we need it described in the docstring.

Reply:
yep they go through '/etc/hosts'  the pure go one use [readHost](https://github.com/golang/go/blob/e2b620ce730992c7991951b6b2d005f70e8c8d36/src/net/hosts.go#L50) where the `testHookHostsPath` is `/etc/hosts` and the cgo implementations all use the libc resolver which will use `/etc/hosts` if your system is setup to do that, and if it isn't setup that way you don't want to parse `/etc/hosts` anyway.